File: Python/Leetcode Easy/637AverageOfLevelsInBinaryTree.py
#https://leetcode.com/problems/average-of-levels-in-binary-tree/description/
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def averageOfLevels(self, root: Optional[TreeNode]) -> List[float]:
        levels = []
        def helper(node, level):
            if node == None:
                return
            if len(levels) == level:
                levels.append([node.val, 1])
            else:
                levels[level][0] += node.val
                levels[level][1] += 1
            helper(node.left, level+1)
            helper(node.right, level+1)
        helper(root, 0)
        average = [0 for __ in range(len(levels))]
        for i, value in enumerate(levels):
            average[i] = value[0] / value[1]
        return average

File: C++/Leetcode Easy/637AverageOfLevelsInBinaryTree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
#include <list>
using namespace std;
class Solution {
public:
    vector<double> averageOfLevels(TreeNode* root) {
        vector<double> res;
        list<TreeNode*> queue = {root};
        while (!queue.empty()) {
            int n = queue.size();
            double average = 0;
            for (int i = 0; i < n; ++i) {
                TreeNode* curr = queue.front();
                if (curr->left) queue.push_back(curr->left);
                if (curr->right) queue.push_back(curr->right);
                queue.pop_front();
                average += curr->val;
            }
            res.push_back(average/n);
        }
        return res;
    }
};